prov/tcp/src/tcpx.h

```diff
@@ -282,10 +285,67 @@ struct tcpx_ep {
 	bool			pollout_set;
 };
 
+struct tcpx_progress {
+	struct fid		fid;
+	ofi_mutex_t		lock;
+	struct dlist_entry	ep_list;
+	struct dlist_entry	active_wait_list;
+
+	struct fd_signal	signal;
+	enum fi_wait_obj	wait_obj;
+	union {
+		ofi_epoll_t	epoll_fd;
+		struct ofi_pollfds *pollfds;
+	};
+
+	pthread_t		thread;
+	bool			auto_progress;
+};
+
+int tcpx_init_progress(struct tcpx_progress *progress);
+void tcpx_close_progress(struct tcpx_progress *progress);
+int tcpx_start_progress(struct tcpx_progress *progress);
+void tcpx_stop_progress(struct tcpx_progress *progress);
+
+void tcpx_run_progress(struct tcpx_progress *progress);
+void tcpx_run_ep(struct tcpx_ep *ep, bool pin, bool pout, bool perr);
+void tcpx_run_pep(struct tcpx_pep *pep, bool pin, bool pout, bool perr);
+void tcpx_run_conn(struct tcpx_conn_handle *conn, bool pin, bool pout, bool perr);
+
+int tcpx_trywait(struct tcpx_progress *progress);
+void tcpx_update_poll(struct tcpx_ep *ep);
+
+int tcpx_monitor_ep(struct tcpx_ep *ep, uint32_t events);
+void tcpx_halt_ep(struct tcpx_ep *ep);
+int tcpx_monitor_pep(struct tcpx_pep *pep);
+void tcpx_halt_pep(struct tcpx_pep *pep);
+int tcpx_monitor_conn(struct tcpx_conn_handle *conn);
+void tcpx_halt_conn(struct tcpx_conn_handle *conn);
+
+
 struct tcpx_fabric {
 	struct util_fabric	util_fabric;
+	struct tcpx_progress	progress;
```

Comment:
Say I have a multi-threaded application where each thread runs with its own collection of EPs, EQs and CQs, all belonging to the same fabric (operations accessing the same resources are serialized on the same thread).

What will happen if all the threads concurrently run `fi_cq_read()`? Will they all compete for the same progress lock? If so, what do you suggest for multi-threaded applications that want to run lockless? Do they need to create one fabric object per thread?

Reply:
Currently, the progress engine is associated with the fabric object.  As long as 2 CQs fall under the same fabric, they will compete for the progress lock.  As the patches are structured, each thread would need to create its own fabric object to run lockless (eventually).

I looked at associating the progress engine with the domain.  The issue is how to handle listening sockets.  Another scaling issue I've seen is slowness setting up new connections, so I want to be able to monitor new connections for faster setup.  My plan is to follow these patches with a 'fast connect' option that rxm can use.  This would allow tcp to invoke a callback after accepting a new socket, so the rxm CM messages can be sent immediately, without needing to write and process events on the EQ.  (When there's a 1000 active connections all sending and receiving data, it can take the app a long time to get around to checking the EQ to setup a new connection.  And most new connections require that the server process 2 EQ events before the EP is usable.)

There are a couple other options for how to handle listening sockets, but this seemed the easiest approach for now.